File: server/lib/langchainChain.js
const { ChatGoogleGenerativeAI } = require("@langchain/google-genai");
const { ChatPromptTemplate } = require("@langchain/core/prompts");
const { StructuredOutputParser } = require("@langchain/core/output_parsers");
const { RunnableSequence } = require("@langchain/core/runnables");
const { z } = require("zod");

const claimSchema = z.object({
  policyNumber: z.string().nullable(),
  policyholderName: z.string().nullable(),
  effectiveStartDate: z.string().nullable(),
  effectiveEndDate: z.string().nullable(),
  incidentDate: z.string().nullable(),
  incidentTime: z.string().nullable(),
  incidentLocation: z.string().nullable(),
  incidentDescription: z.string().nullable(),
  claimantName: z.string().nullable(),
  thirdParties: z.array(z.string()),
  contactDetails: z.string().nullable(),
  assetType: z.string().nullable(),
  assetId: z.string().nullable(),
  estimatedDamage: z.number().nullable(),
  claimType: z.string().nullable(),
  initialEstimate: z.number().nullable(),
});

const parser = StructuredOutputParser.fromZodSchema(claimSchema);

const prompt = ChatPromptTemplate.fromMessages([
  [
    "system",
    "You are an insurance FNOL data extraction expert. Extract all fields from the document. Return ONLY valid JSON matching the schema. {format_instructions}",
  ],
  ["human", "Extract fields from this FNOL document:\n\n{document}"],
]);

let chain = null;

function getChain() {
  if (!chain) {
    const model = new ChatGoogleGenerativeAI({
      model: "gemini-2.5-flash",
      apiKey: process.env.GOOGLE_API_KEY,
      temperature: 0,
    });
    chain = RunnableSequence.from([prompt, model, parser]);
  }
  return chain;
}

async function extractClaimFields(rawText) {
  return await getChain().invoke({
    document: rawText,
    format_instructions: parser.getFormatInstructions(),
  });
}

module.exports = { extractClaimFields };

File: server/lib/langchain_chain.py
import os
from typing import Optional
from pydantic import BaseModel
from langchain_google_genai import ChatGoogleGenerativeAI
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers import PydanticOutputParser
from langchain_core.runnables import RunnableSequence


class ClaimSchema(BaseModel):
    policyNumber: Optional[str] = None
    policyholderName: Optional[str] = None
    effectiveStartDate: Optional[str] = None
    effectiveEndDate: Optional[str] = None
    incidentDate: Optional[str] = None
    incidentTime: Optional[str] = None
    incidentLocation: Optional[str] = None
    incidentDescription: Optional[str] = None
    claimantName: Optional[str] = None
    thirdParties: list[str] = []
    contactDetails: Optional[str] = None
    assetType: Optional[str] = None
    assetId: Optional[str] = None
    estimatedDamage: Optional[float] = None
    claimType: Optional[str] = None
    initialEstimate: Optional[float] = None


parser = PydanticOutputParser(pydantic_object=ClaimSchema)

prompt = ChatPromptTemplate.from_messages([
    ("system",
     "You are an insurance FNOL data extraction expert. Extract all fields from the document. "
     "Return ONLY valid JSON matching the schema. {format_instructions}"),
    ("human", "Extract fields from this FNOL document:\n\n{document}"),
])

_chain = None


def _get_chain() -> RunnableSequence:
    global _chain
    if _chain is None:
        model = ChatGoogleGenerativeAI(
            model="gemini-2.5-flash",
            api_key=os.getenv("GOOGLE_API_KEY"),
            temperature=0,
        )
        _chain = prompt | model | parser
    return _chain


async def extract_claim_fields(raw_text: str) -> ClaimSchema:
    chain = _get_chain()
    result = await chain.ainvoke({
        "document": raw_text,
        "format_instructions": parser.get_format_instructions(),
    })
    return result
